File: aihound/scanners/docker.py
"""Scanner for Docker credentials in ~/.docker/config.json."""

from __future__ import annotations

import json
import logging
from pathlib import Path

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType, RiskLevel,
)
from aihound.core.platform import detect_platform, Platform, get_home, get_wsl_windows_home
from aihound.core.redactor import mask_value
from aihound.core.permissions import (
    get_file_permissions, get_file_owner, assess_risk,
    get_file_mtime, describe_staleness,
)
from aihound.remediation import hint_use_credential_helper
from aihound.scanners import register

logger = logging.getLogger("aihound.scanners.docker")


@register
class DockerScanner(BaseScanner):
    def name(self) -> str:
        return "Docker"

    def slug(self) -> str:
        return "docker"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        for path in self._get_config_paths(plat):
            self._scan_docker_config(path, result, show_secrets)

        return result

    def _get_config_paths(self, plat: Platform) -> list[Path]:
        paths: list[Path] = []
        home = get_home()

        # Docker stores config in ~/.docker/config.json on all platforms
        # (on Windows, that resolves under %USERPROFILE%/.docker/config.json)
        paths.append(home / ".docker" / "config.json")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".docker" / "config.json")

        return paths

    def _scan_docker_config(
        self, path: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not path.exists():
            logger.debug("Docker config not found: %s", path)
            return

        logger.debug("Reading docker config: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.warning("Failed to parse %s: %s", path, e, exc_info=True)
            result.errors.append(f"Failed to parse {path}: {e}")
            return

        if not isinstance(data, dict):
            return

        storage = StorageType.PLAINTEXT_JSON
        staleness_note = f"File last modified: {describe_staleness(mtime)}" if mtime else None

        # 1) auths dict: { "registry": { "auth": "base64(user:pass)", ... } }
        auths = data.get("auths")
        if isinstance(auths, dict):
            for registry, entry in auths.items():
                if not isinstance(entry, dict):
                    continue

                auth_b64 = entry.get("auth")
                if isinstance(auth_b64, str) and auth_b64:
                    notes = [f"Registry: {registry}", "Base64(user:password) stored in plaintext"]
                    if staleness_note:
                        notes.append(staleness_note)
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=f"registry_auth:{registry}",
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(auth_b64, show_full=show_secrets),
                        raw_value=auth_b64 if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation=(
                            "Use docker credential helpers (credsStore) instead of "
                            "storing tokens in config.json. See: docker login --help"
                        ),
                        remediation_hint=hint_use_credential_helper(
                            "docker", ["osxkeychain", "pass", "secretservice"]
                        ),
                        notes=notes,
                    ))

                # Sometimes an identitytoken is stored alongside auth
                id_token = entry.get("identitytoken")
                if isinstance(id_token, str) and id_token:
                    notes = [f"Registry: {registry}", "Docker identity token (OAuth refresh-like)"]
                    if staleness_note:
                        notes.append(staleness_note)
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=f"registry_identitytoken:{registry}",
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(id_token, show_full=show_secrets),
                        raw_value=id_token if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation=(
                            "Use docker credential helpers (credsStore) instead of "
                            "storing tokens in config.json. See: docker login --help"
                        ),
                        remediation_hint=hint_use_credential_helper(
                            "docker", ["osxkeychain", "pass", "secretservice"]
                        ),
                        notes=notes,
                    ))

                # Any other string values that look like a secret key
                for sub_key, sub_val in entry.items():
                    if sub_key in ("auth", "identitytoken", "email", "username"):
                        continue
                    if isinstance(sub_val, str) and len(sub_val) > 20:
                        lowered = sub_key.lower()
                        if any(k in lowered for k in ("token", "secret", "key", "password")):
                            notes = [f"Registry: {registry}", f"Field: {sub_key}"]
                            if staleness_note:
                                notes.append(staleness_note)
                            result.findings.append(CredentialFinding(
                                tool_name=self.name(),
                                credential_type=f"registry_{sub_key}:{registry}",
                                storage_type=storage,
                                location=str(path),
                                exists=True,
                                risk_level=assess_risk(storage, path),
                                value_preview=mask_value(sub_val, show_full=show_secrets),
                                raw_value=sub_val if show_secrets else None,
                                file_permissions=perms,
                                file_owner=owner,
                                file_modified=mtime,
                                remediation=(
                                    "Use docker credential helpers (credsStore) instead of "
                                    "storing tokens in config.json. See: docker login --help"
                                ),
                                remediation_hint=hint_use_credential_helper(
                                    "docker", ["osxkeychain", "pass", "secretservice"]
                                ),
                                notes=notes,
                            ))

        # 2) credsStore: global credential helper (safer)
        creds_store = data.get("credsStore")
        if isinstance(creds_store, str) and creds_store:
            notes = [
                f"Using credential helper: docker-credential-{creds_store}",
                "Credentials stored outside config.json (likely in OS keystore)",
            ]
            if staleness_note:
                notes.append(staleness_note)
            result.findings.append(CredentialFinding(
                tool_name=self.name(),
                credential_type="credsStore",
                storage_type=StorageType.UNKNOWN,
                location=str(path),
                exists=True,
                risk_level=RiskLevel.INFO,
                value_preview=creds_store,
                file_permissions=perms,
                file_owner=owner,
                file_modified=mtime,
                remediation=(
                    "Use docker credential helpers (credsStore) instead of "
                    "storing tokens in config.json. See: docker login --help"
                ),
                remediation_hint=hint_use_credential_helper(
                    "docker", ["osxkeychain", "pass", "secretservice"]
                ),
                notes=notes,
            ))

        # 3) credHelpers: per-registry credential helpers
        cred_helpers = data.get("credHelpers")
        if isinstance(cred_helpers, dict) and cred_helpers:
            for registry, helper in cred_helpers.items():
                if not isinstance(helper, str):
                    continue
                notes = [
                    f"Registry: {registry}",
                    f"Using credential helper: docker-credential-{helper}",
                    "Credentials stored outside config.json (likely in OS keystore)",
                ]
                if staleness_note:
                    notes.append(staleness_note)
                result.findings.append(CredentialFinding(
                    tool_name=self.name(),
                    credential_type=f"credHelper:{registry}",
                    storage_type=StorageType.UNKNOWN,
                    location=str(path),
                    exists=True,
                    risk_level=RiskLevel.INFO,
                    value_preview=helper,
                    file_permissions=perms,
                    file_owner=owner,
                    file_modified=mtime,
                    remediation=(
                        "Use docker credential helpers (credsStore) instead of "
                        "storing tokens in config.json. See: docker login --help"
                    ),
                    remediation_hint=hint_use_credential_helper(
                        "docker", ["osxkeychain", "pass", "secretservice"]
                    ),
                    notes=notes,
                ))

        # 4) Top-level secret-looking keys (e.g., secretName, proxies with tokens)
        for key, val in data.items():
            if key in ("auths", "credsStore", "credHelpers"):
                continue
            if isinstance(val, str) and len(val) > 20:
                lowered = key.lower()
                if any(k in lowered for k in ("token", "secret", "key", "password", "auth")):
                    notes = [f"Top-level field: {key}"]
                    if staleness_note:
                        notes.append(staleness_note)
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=f"config:{key}",
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(val, show_full=show_secrets),
                        raw_value=val if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation=(
                            "Use docker credential helpers (credsStore) instead of "
                            "storing tokens in config.json. See: docker login --help"
                        ),
                        remediation_hint=hint_use_credential_helper(
                            "docker", ["osxkeychain", "pass", "secretservice"]
                        ),
                        notes=notes,
                    ))
            elif isinstance(val, dict):
                # Recurse one level (e.g., "proxies", "HttpHeaders")
                self._recurse_for_secrets(
                    val, path, perms, owner, mtime, result, show_secrets,
                    prefix=key, staleness_note=staleness_note,
                )

    def _recurse_for_secrets(
        self, data: dict, path: Path, perms, owner, mtime,
        result: ScanResult, show_secrets: bool,
        prefix: str, staleness_note: str | None,
        depth: int = 0,
    ) -> None:
        if depth > 4:
            return
        storage = StorageType.PLAINTEXT_JSON
        for key, val in data.items():
            full_key = f"{prefix}.{key}"
            if isinstance(val, dict):
                self._recurse_for_secrets(
                    val, path, perms, owner, mtime, result, show_secrets,
                    prefix=full_key, staleness_note=staleness_note, depth=depth + 1,
                )
            elif isinstance(val, str) and len(val) > 20:
                lowered = key.lower()
                if any(k in lowered for k in ("token", "secret", "key", "password", "auth")):
                    notes = [f"Nested field: {full_key}"]
                    if staleness_note:
                        notes.append(staleness_note)
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=f"config:{full_key}",
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(val, show_full=show_secrets),
                        raw_value=val if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation=(
                            "Use docker credential helpers (credsStore) instead of "
                            "storing tokens in config.json. See: docker login --help"
                        ),
                        remediation_hint=hint_use_credential_helper(
                            "docker", ["osxkeychain", "pass", "secretservice"]
                        ),
                        notes=notes,
                    ))

File: Other Versions/Go/scanners/docker.go
package scanners

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

var dockerSecretKeyTokens = []string{"token", "secret", "key", "password", "auth"}
var dockerRegistrySubKeyTokens = []string{"token", "secret", "key", "password"}

const dockerRemediation = "Use docker credential helpers (credsStore) instead of storing tokens in config.json. See: docker login --help"

type dockerScanner struct{}

func init() {
	Register(&dockerScanner{})
}

func (s *dockerScanner) Name() string        { return "Docker" }
func (s *dockerScanner) Slug() string         { return "docker" }
func (s *dockerScanner) IsApplicable() bool   { return true }

func (s *dockerScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, p := range s.getConfigPaths(plat) {
		s.scanDockerConfig(p, &result, showSecrets)
	}

	return result
}

func (s *dockerScanner) getConfigPaths(plat core.Platform) []string {
	var paths []string
	home := core.GetHome()
	paths = append(paths, filepath.Join(home, ".docker", "config.json"))

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".docker", "config.json"))
		}
	}
	return paths
}

func (s *dockerScanner) scanDockerConfig(path string, result *core.ScanResult, showSecrets bool) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	var data map[string]interface{}
	if err := json.Unmarshal(raw, &data); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Failed to parse %s: %v", path, err))
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	mtime := core.GetFileMtime(path)
	storage := core.PlaintextJSON

	stalenessNote := ""
	if mt := core.GetFileMtimeTime(path); !mt.IsZero() {
		stalenessNote = "File last modified: " + core.DescribeStaleness(mt)
	}

	// 1) auths dict
	if authsRaw, ok := data["auths"]; ok {
		if auths, ok := authsRaw.(map[string]interface{}); ok {
			for registry, entryRaw := range auths {
				entry, ok := entryRaw.(map[string]interface{})
				if !ok {
					continue
				}

				if authB64, ok := entry["auth"].(string); ok && authB64 != "" {
					notes := []string{
						fmt.Sprintf("Registry: %s", registry),
						"Base64(user:password) stored in plaintext",
					}
					if stalenessNote != "" {
						notes = append(notes, stalenessNote)
					}
					rawValue := ""
					if showSecrets {
						rawValue = authB64
					}
					result.Findings = append(result.Findings, core.CredentialFinding{
						ToolName:        s.Name(),
						CredentialType:  fmt.Sprintf("registry_auth:%s", registry),
						StorageType:     storage,
						Location:        path,
						Exists:          true,
						RiskLevel:       core.AssessRisk(storage, path),
						ValuePreview:    core.MaskValue(authB64, showSecrets),
						RawValue:        rawValue,
						FilePermissions: perms,
						FileOwner:       owner,
						FileModified:    mtime,
						Remediation:     dockerRemediation,
						RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
						Notes:           notes,
					})
				}

				if idToken, ok := entry["identitytoken"].(string); ok && idToken != "" {
					notes := []string{
						fmt.Sprintf("Registry: %s", registry),
						"Docker identity token (OAuth refresh-like)",
					}
					if stalenessNote != "" {
						notes = append(notes, stalenessNote)
					}
					rawValue := ""
					if showSecrets {
						rawValue = idToken
					}
					result.Findings = append(result.Findings, core.CredentialFinding{
						ToolName:        s.Name(),
						CredentialType:  fmt.Sprintf("registry_identitytoken:%s", registry),
						StorageType:     storage,
						Location:        path,
						Exists:          true,
						RiskLevel:       core.AssessRisk(storage, path),
						ValuePreview:    core.MaskValue(idToken, showSecrets),
						RawValue:        rawValue,
						FilePermissions: perms,
						FileOwner:       owner,
						FileModified:    mtime,
						Remediation:     dockerRemediation,
						RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
						Notes:           notes,
					})
				}

				for subKey, subValRaw := range entry {
					if subKey == "auth" || subKey == "identitytoken" || subKey == "email" || subKey == "username" {
						continue
					}
					subVal, ok := subValRaw.(string)
					if !ok || len(subVal) <= 20 {
						continue
					}
					lowered := strings.ToLower(subKey)
					match := false
					for _, k := range dockerRegistrySubKeyTokens {
						if strings.Contains(lowered, k) {
							match = true
							break
						}
					}
					if !match {
						continue
					}

					notes := []string{
						fmt.Sprintf("Registry: %s", registry),
						fmt.Sprintf("Field: %s", subKey),
					}
					if stalenessNote != "" {
						notes = append(notes, stalenessNote)
					}
					rawValue := ""
					if showSecrets {
						rawValue = subVal
					}
					result.Findings = append(result.Findings, core.CredentialFinding{
						ToolName:        s.Name(),
						CredentialType:  fmt.Sprintf("registry_%s:%s", subKey, registry),
						StorageType:     storage,
						Location:        path,
						Exists:          true,
						RiskLevel:       core.AssessRisk(storage, path),
						ValuePreview:    core.MaskValue(subVal, showSecrets),
						RawValue:        rawValue,
						FilePermissions: perms,
						FileOwner:       owner,
						FileModified:    mtime,
						Remediation:     dockerRemediation,
						RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
						Notes:           notes,
					})
				}
			}
		}
	}

	// 2) credsStore
	if credsStore, ok := data["credsStore"].(string); ok && credsStore != "" {
		notes := []string{
			fmt.Sprintf("Using credential helper: docker-credential-%s", credsStore),
			"Credentials stored outside config.json (likely in OS keystore)",
		}
		if stalenessNote != "" {
			notes = append(notes, stalenessNote)
		}
		result.Findings = append(result.Findings, core.CredentialFinding{
			ToolName:        s.Name(),
			CredentialType:  "credsStore",
			StorageType:     core.StorageUnknown,
			Location:        path,
			Exists:          true,
			RiskLevel:       core.RiskInfo,
			ValuePreview:    credsStore,
			FilePermissions: perms,
			FileOwner:       owner,
			FileModified:    mtime,
			Remediation:     dockerRemediation,
			RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
			Notes:           notes,
		})
	}

	// 3) credHelpers
	if credHelpersRaw, ok := data["credHelpers"]; ok {
		if credHelpers, ok := credHelpersRaw.(map[string]interface{}); ok {
			for registry, helperRaw := range credHelpers {
				helper, ok := helperRaw.(string)
				if !ok {
					continue
				}
				notes := []string{
					fmt.Sprintf("Registry: %s", registry),
					fmt.Sprintf("Using credential helper: docker-credential-%s", helper),
					"Credentials stored outside config.json (likely in OS keystore)",
				}
				if stalenessNote != "" {
					notes = append(notes, stalenessNote)
				}
				result.Findings = append(result.Findings, core.CredentialFinding{
					ToolName:        s.Name(),
					CredentialType:  fmt.Sprintf("credHelper:%s", registry),
					StorageType:     core.StorageUnknown,
					Location:        path,
					Exists:          true,
					RiskLevel:       core.RiskInfo,
					ValuePreview:    helper,
					FilePermissions: perms,
					FileOwner:       owner,
					FileModified:    mtime,
					Remediation:     dockerRemediation,
					RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
					Notes:           notes,
				})
			}
		}
	}

	// 4) Top-level secret-looking keys / recurse one level
	for key, val := range data {
		if key == "auths" || key == "credsStore" || key == "credHelpers" {
			continue
		}
		switch v := val.(type) {
		case string:
			if len(v) <= 20 {
				continue
			}
			lowered := strings.ToLower(key)
			match := false
			for _, k := range dockerSecretKeyTokens {
				if strings.Contains(lowered, k) {
					match = true
					break
				}
			}
			if !match {
				continue
			}
			notes := []string{fmt.Sprintf("Top-level field: %s", key)}
			if stalenessNote != "" {
				notes = append(notes, stalenessNote)
			}
			rawValue := ""
			if showSecrets {
				rawValue = v
			}
			result.Findings = append(result.Findings, core.CredentialFinding{
				ToolName:        s.Name(),
				CredentialType:  fmt.Sprintf("config:%s", key),
				StorageType:     storage,
				Location:        path,
				Exists:          true,
				RiskLevel:       core.AssessRisk(storage, path),
				ValuePreview:    core.MaskValue(v, showSecrets),
				RawValue:        rawValue,
				FilePermissions: perms,
				FileOwner:       owner,
				FileModified:    mtime,
				Remediation:     dockerRemediation,
				RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
				Notes:           notes,
			})
		case map[string]interface{}:
			s.recurseForSecrets(v, path, perms, owner, mtime, result, showSecrets, key, stalenessNote, 0)
		}
	}
}

func (s *dockerScanner) recurseForSecrets(data map[string]interface{}, path, perms, owner, mtime string, result *core.ScanResult, showSecrets bool, prefix, stalenessNote string, depth int) {
	if depth > 4 {
		return
	}
	storage := core.PlaintextJSON
	for key, val := range data {
		fullKey := prefix + "." + key
		switch v := val.(type) {
		case map[string]interface{}:
			s.recurseForSecrets(v, path, perms, owner, mtime, result, showSecrets, fullKey, stalenessNote, depth+1)
		case string:
			if len(v) <= 20 {
				continue
			}
			lowered := strings.ToLower(key)
			match := false
			for _, k := range dockerSecretKeyTokens {
				if strings.Contains(lowered, k) {
					match = true
					break
				}
			}
			if !match {
				continue
			}
			notes := []string{fmt.Sprintf("Nested field: %s", fullKey)}
			if stalenessNote != "" {
				notes = append(notes, stalenessNote)
			}
			rawValue := ""
			if showSecrets {
				rawValue = v
			}
			result.Findings = append(result.Findings, core.CredentialFinding{
				ToolName:        s.Name(),
				CredentialType:  fmt.Sprintf("config:%s", fullKey),
				StorageType:     storage,
				Location:        path,
				Exists:          true,
				RiskLevel:       core.AssessRisk(storage, path),
				ValuePreview:    core.MaskValue(v, showSecrets),
				RawValue:        rawValue,
				FilePermissions: perms,
				FileOwner:       owner,
				FileModified:    mtime,
				Remediation:     dockerRemediation,
				RemediationHint: remediation.HintUseCredentialHelper("docker", []string{"osxkeychain", "pass", "secretservice"}),
				Notes:           notes,
			})
		}
	}
}
